File: services/analytics/app/formulas.py
"""Formula engine per analisi investimenti immobiliari SRL.

Tutti gli importi usano Decimal. Nessuna consulenza fiscale: parametri configurabili.
"""

from decimal import Decimal, ROUND_HALF_UP
from typing import Any

from app.schemas import (
    AnalysisRequest,
    DealStrategy,
    ScenarioResult,
    SensitivitySignal,
    TaxRegime,
)


TWOPLACES = Decimal("0.01")
FOURPLACES = Decimal("0.0001")


def _round_money(value: Decimal) -> Decimal:
    return value.quantize(TWOPLACES, rounding=ROUND_HALF_UP)


def _round_rate(value: Decimal) -> Decimal:
    return value.quantize(FOURPLACES, rounding=ROUND_HALF_UP)


def acquisition_price(request: AnalysisRequest) -> Decimal:
    discount = request.acquisition.target_discount_pct * request.scenario_multiplier
    if request.scenario_multiplier > Decimal("1"):
        discount = request.acquisition.target_discount_pct * (
            Decimal("2") - request.scenario_multiplier
        )
    price = request.acquisition.asking_price * (Decimal("1") - discount)
    return _round_money(max(price, Decimal("0")))


def notary_and_acquisition_costs(purchase_price: Decimal, request: AnalysisRequest) -> Decimal:
    pct = request.acquisition.notary_and_fees_pct
    if request.tax_profile.tax_regime == TaxRegime.REGISTRY:
        reg_tax = purchase_price * request.tax_profile.registration_tax_rate
        return _round_money(purchase_price * pct + reg_tax)
    vat = purchase_price * request.tax_profile.vat_rate
    return _round_money(purchase_price * pct + vat)


def renovation_total(request: AnalysisRequest) -> Decimal:
  base = request.renovation.total_capex
  if request.scenario_multiplier > Decimal("1"):
      base = base * request.scenario_multiplier
  elif request.scenario_multiplier < Decimal("1"):
      base = base * (Decimal("2") - request.scenario_multiplier)
  contingency = base * request.renovation.contingency_pct * request.scenario_multiplier
  return _round_money(base + contingency)


def total_project_cost(purchase: Decimal, acquisition_costs: Decimal, capex: Decimal) -> Decimal:
    return _round_money(purchase + acquisition_costs + capex)


def initial_capital(
    total_cost: Decimal,
    loan_amount: Decimal,
) -> Decimal:
    return _round_money(max(total_cost - loan_amount, Decimal("0")))


def compute_ltv(loan_amount: Decimal, property_value: Decimal) -> Decimal | None:
    if property_value <= 0:
        return None
    return _round_rate(loan_amount / property_value)


def annual_debt_service(
    loan_amount: Decimal,
    rate_annual: Decimal,
    term_years: int,
) -> Decimal:
    if loan_amount <= 0:
        return Decimal("0")
    monthly_rate = rate_annual / Decimal("12")
    n = term_years * 12
    if monthly_rate == 0:
        return _round_money(loan_amount / Decimal(n) * Decimal("12"))
    factor = (Decimal("1") + monthly_rate) ** n
    monthly_payment = loan_amount * monthly_rate * factor / (factor - Decimal("1"))
    return _round_money(monthly_payment * Decimal("12"))


def compute_dscr(
    net_operating_income: Decimal,
    debt_service: Decimal,
) -> Decimal | None:
    if debt_service <= 0:
        return None
    return _round_rate(net_operating_income / debt_service)


def corporate_tax(amount: Decimal, request: AnalysisRequest) -> Decimal:
    if amount <= 0:
        return Decimal("0")
    ires = amount * request.tax_profile.ires_rate
    irap = amount * request.tax_profile.irap_rate
    return _round_money(ires + irap)


def annual_net_rental_income(request: AnalysisRequest) -> Decimal:
    gross = request.rental.monthly_rent * Decimal("12")
    vacancy = gross * request.rental.vacancy_rate
    maintenance = gross * request.rental.annual_maintenance_pct
    management = gross * request.rental.property_management_pct
    net = gross - vacancy - maintenance - management
    return _round_money(max(net, Decimal("0")))


def sale_margins(
    sale_price: Decimal,
    total_cost: Decimal,
    request: AnalysisRequest,
) -> tuple[Decimal, Decimal]:
    sale_costs = sale_price * request.sale.sale_costs_pct
    gross = _round_money(sale_price - total_cost - sale_costs)
    tax = corporate_tax(gross, request) if gross > 0 else Decimal("0")
    net = _round_money(gross - tax)
    return gross, net


def compute_npv(
    cash_flows: list[Decimal],
    discount_rate_annual: Decimal,
) -> Decimal:
    monthly_rate = discount_rate_annual / Decimal("12")
    npv = Decimal("0")
    for month, cf in enumerate(cash_flows):
        discount = (Decimal("1") + monthly_rate) ** month
        npv += cf / discount
    return _round_money(npv)


def compute_irr(cash_flows: list[Decimal], max_iterations: int = 100) -> Decimal | None:
    if not cash_flows or cash_flows[0] >= 0:
        return None

    low = Decimal("-0.99")
    high = Decimal("5")
    for _ in range(max_iterations):
        mid = (low + high) / Decimal("2")
        npv = Decimal("0")
        for t, cf in enumerate(cash_flows):
            npv += cf / (Decimal("1") + mid) ** t
        if abs(npv) < Decimal("0.01"):
            return _round_rate(mid * Decimal("12"))
        if npv > 0:
            low = mid
        else:
            high = mid
    return _round_rate(mid * Decimal("12"))


def build_cash_flows(
    request: AnalysisRequest,
    purchase: Decimal,
    acquisition_costs: Decimal,
    capex: Decimal,
    strategy: DealStrategy,
) -> list[Decimal]:
    months = max(request.timeline.exit_month, request.sale.holding_months, 12)
    flows: list[Decimal] = [Decimal("0")] * (months + 1)

    equity_out = initial_capital(
        total_project_cost(purchase, acquisition_costs, capex),
        request.financing.loan_amount,
    )
    flows[0] = -equity_out
    flows[request.timeline.acquisition_month] -= purchase + acquisition_costs

    reno_start = request.timeline.renovation_start_month
    reno_months = max(request.renovation.duration_months, 1)
    capex_monthly = capex / Decimal(reno_months)
    for m in range(reno_start, min(reno_start + reno_months, months + 1)):
        flows[m] -= capex_monthly

    debt_monthly = annual_debt_service(
        request.financing.loan_amount,
        request.financing.interest_rate_annual,
        request.financing.loan_term_years,
    ) / Decimal("12")

    if strategy in (DealStrategy.BUY_RENOVATE_RENT, DealStrategy.BUY_HOLD_SELL):
        net_rent = annual_net_rental_income(request) / Decimal("12")
        for m in range(reno_start + reno_months, months + 1):
            flows[m] += net_rent - debt_monthly
    else:
        for m in range(1, months + 1):
            flows[m] -= debt_monthly

    if strategy in (DealStrategy.FIX_FLIP, DealStrategy.BUY_HOLD_SELL):
        exit_m = min(request.timeline.exit_month, months)
        sale_price = request.sale.expected_sale_price
        if request.scenario_multiplier < Decimal("1"):
            sale_price = sale_price * request.scenario_multiplier
        elif request.scenario_multiplier > Decimal("1"):
            sale_price = sale_price * (Decimal("2") - request.scenario_multiplier)
        sale_costs = sale_price * request.sale.sale_costs_pct
        loan_payoff = request.financing.loan_amount
        flows[exit_m] += sale_price - sale_costs - loan_payoff

    return [_round_money(f) for f in flows]


def sensitivity_signal(
    net_margin: Decimal | None,
    dscr: Decimal | None,
    irr: Decimal | None,
    scenario_multiplier: Decimal,
) -> SensitivitySignal:
    score = 0
    if net_margin is not None:
        if net_margin > Decimal("50000"):
            score += 2
        elif net_margin > Decimal("15000"):
            score += 1
        elif net_margin < 0:
            score -= 2
    if dscr is not None:
        if dscr >= Decimal("1.25"):
            score += 1
        elif dscr < Decimal("1"):
            score -= 2
    if irr is not None:
        if irr >= Decimal("0.15"):
            score += 1
        elif irr < Decimal("0.05"):
            score -= 1
    if scenario_multiplier > Decimal("1"):
        score -= 1
    if score >= 2:
        return SensitivitySignal.GREEN
    if score >= 0:
        return SensitivitySignal.AMBER
    return SensitivitySignal.RED


def run_scenario(request: AnalysisRequest) -> ScenarioResult:
    purchase = acquisition_price(request)
    acq_costs = notary_and_acquisition_costs(purchase, request)
    capex = renovation_total(request)
    total_cost = total_project_cost(purchase, acq_costs, capex)
    equity = initial_capital(total_cost, request.financing.loan_amount)

    prop_value = request.financing.property_value_for_ltv or (
        request.sale.expected_sale_price or purchase
    )
    ltv = compute_ltv(request.financing.loan_amount, prop_value)

    gross_margin: Decimal | None = None
    net_margin: Decimal | None = None
    if request.strategy in (DealStrategy.FIX_FLIP, DealStrategy.BUY_HOLD_SELL):
        sale_price = request.sale.expected_sale_price
        if request.scenario_multiplier < Decimal("1"):
            sale_price = sale_price * request.scenario_multiplier
        gross_margin, net_margin = sale_margins(sale_price, total_cost, request)

    net_rent = None
    monthly_cf = None
    if request.strategy in (DealStrategy.BUY_RENOVATE_RENT, DealStrategy.BUY_HOLD_SELL):
        net_rent = annual_net_rental_income(request)
        debt = annual_debt_service(
            request.financing.loan_amount,
            request.financing.interest_rate_annual,
            request.financing.loan_term_years,
        )
        monthly_cf = _round_money((net_rent - debt) / Decimal("12"))

    debt_service = annual_debt_service(
        request.financing.loan_amount,
        request.financing.interest_rate_annual,
        request.financing.loan_term_years,
    )
    dscr = compute_dscr(net_rent or Decimal("0"), debt_service)

    flows = build_cash_flows(request, purchase, acq_costs, capex, request.strategy)
    npv = compute_npv(flows, request.timeline.discount_rate_annual)
    irr = compute_irr(flows)

    signal = sensitivity_signal(net_margin, dscr, irr, request.scenario_multiplier)

    assumptions: dict[str, Any] = {
        "purchase_price": str(purchase),
        "acquisition_costs": str(acq_costs),
        "capex": str(capex),
        "total_cost": str(total_cost),
        "scenario_multiplier": str(request.scenario_multiplier),
    }

    return ScenarioResult(
        initial_capital_required=equity,
        total_project_cost=total_cost,
        gross_sale_margin=gross_margin,
        net_sale_margin=net_margin,
        annual_net_rental_income=net_rent,
        monthly_cash_flow=monthly_cf,
        ltv=ltv,
        dscr=dscr,
        npv=npv,
        irr=irr,
        sensitivity_signal=signal,
        assumptions_used=assumptions,
    )


def run_full_analysis(base_request: AnalysisRequest):
    from app.schemas import AnalysisResponse

    base = run_scenario(base_request.model_copy(update={"scenario_multiplier": Decimal("1")}))
    prudent = run_scenario(
        base_request.model_copy(update={"scenario_multiplier": Decimal("0.85")})
    )
    stress = run_scenario(
        base_request.model_copy(update={"scenario_multiplier": Decimal("1.25")})
    )

    signals = [base.sensitivity_signal, prudent.sensitivity_signal, stress.sensitivity_signal]
    if all(s == SensitivitySignal.GREEN for s in signals):
        summary = "Operazione solida su tutti gli scenari. Margine regge anche in caso prudente."
    elif stress.sensitivity_signal == SensitivitySignal.RED:
        summary = (
            "Attenzione: lo scenario stress è in rosso. "
            "Servono maggiore sconto, capex contenuto o tempi più rapidi."
        )
    else:
        summary = (
            "Operazione fattibile nel caso base ma sensibile a capex e tempi. "
            "Valutare negoziazione aggiuntiva."
        )

    return AnalysisResponse(
        base_case=base,
        prudent_case=prudent,
        stress_case=stress,
        sensitivity_summary=summary,
    )

File: apps/web/src/lib/analytics-engine/formulas.ts
import type {
  AnalysisRequestInput,
  AnalysisResponse,
  DealStrategy,
  ScenarioResult,
  SensitivitySignal,
  TaxRegime,
} from "./types";

interface ScenarioFactors {
  discount_factor: number;
  capex_factor: number;
  sale_price_factor: number;
}

interface NormalizedRequest {
  strategy: DealStrategy;
  acquisition: {
    asking_price: number;
    target_discount_pct: number;
    notary_and_fees_pct: number;
  };
  renovation: {
    total_capex: number;
    contingency_pct: number;
    duration_months: number;
  };
  financing: {
    loan_amount: number;
    interest_rate_annual: number;
    loan_term_years: number;
    property_value_for_ltv: number | null;
  };
  rental: {
    monthly_rent: number;
    vacancy_rate: number;
    annual_maintenance_pct: number;
    property_management_pct: number;
  };
  sale: {
    expected_sale_price: number;
    sale_costs_pct: number;
    holding_months: number;
  };
  tax_profile: {
    ires_rate: number;
    irap_rate: number;
    registration_tax_rate: number;
    vat_rate: number;
    tax_regime: TaxRegime;
  };
  timeline: {
    acquisition_month: number;
    renovation_start_month: number;
    exit_month: number;
    discount_rate_annual: number;
  };
  scenario_multiplier: number;
}

function roundMoney(value: number): number {
  return Math.round(value * 100) / 100;
}

function roundRate(value: number): number {
  return Math.round(value * 10000) / 10000;
}

function normalizeStrategy(strategy: string): DealStrategy {
  if (strategy === "buy_renovate_rent" || strategy === "buy_hold_sell") {
    return strategy;
  }
  return "fix_flip";
}

function normalizeRequest(input: AnalysisRequestInput, multiplier?: number): NormalizedRequest {
  return {
    strategy: normalizeStrategy(input.strategy),
    acquisition: {
      asking_price: input.acquisition.asking_price,
      target_discount_pct: input.acquisition.target_discount_pct ?? 0.05,
      notary_and_fees_pct: input.acquisition.notary_and_fees_pct ?? 0.03,
    },
    renovation: {
      total_capex: input.renovation.total_capex,
      contingency_pct: input.renovation.contingency_pct ?? 0.1,
      duration_months: input.renovation.duration_months ?? 6,
    },
    financing: {
      loan_amount: input.financing?.loan_amount ?? 0,
      interest_rate_annual: input.financing?.interest_rate_annual ?? 0.04,
      loan_term_years: input.financing?.loan_term_years ?? 20,
      property_value_for_ltv: input.financing?.property_value_for_ltv ?? null,
    },
    rental: {
      monthly_rent: input.rental?.monthly_rent ?? 0,
      vacancy_rate: input.rental?.vacancy_rate ?? 0.08,
      annual_maintenance_pct: input.rental?.annual_maintenance_pct ?? 0.05,
      property_management_pct: input.rental?.property_management_pct ?? 0.08,
    },
    sale: {
      expected_sale_price: input.sale?.expected_sale_price ?? 0,
      sale_costs_pct: input.sale?.sale_costs_pct ?? 0.04,
      holding_months: input.sale?.holding_months ?? 12,
    },
    tax_profile: {
      ires_rate: input.tax_profile?.ires_rate ?? 0.24,
      irap_rate: input.tax_profile?.irap_rate ?? 0.039,
      registration_tax_rate: input.tax_profile?.registration_tax_rate ?? 0.09,
      vat_rate: input.tax_profile?.vat_rate ?? 0.22,
      tax_regime: input.tax_profile?.tax_regime ?? "registry",
    },
    timeline: {
      acquisition_month: input.timeline?.acquisition_month ?? 0,
      renovation_start_month: input.timeline?.renovation_start_month ?? 1,
      exit_month: input.timeline?.exit_month ?? 12,
      discount_rate_annual: input.timeline?.discount_rate_annual ?? 0.08,
    },
    scenario_multiplier: multiplier ?? input.scenario_multiplier ?? 1,
  };
}

function scenarioFactors(multiplier: number): ScenarioFactors {
  if (multiplier > 1) {
    const stress = 2 - multiplier;
    return {
      discount_factor: stress,
      capex_factor: multiplier,
      sale_price_factor: stress,
    };
  }
  if (multiplier < 1) {
    const prudent = 2 - multiplier;
    return {
      discount_factor: multiplier,
      capex_factor: prudent,
      sale_price_factor: multiplier,
    };
  }
  return { discount_factor: 1, capex_factor: 1, sale_price_factor: 1 };
}

function adjustedSalePrice(expectedSalePrice: number, multiplier: number): number {
  const factors = scenarioFactors(multiplier);
  return roundMoney(expectedSalePrice * factors.sale_price_factor);
}

function acquisitionPrice(request: NormalizedRequest): number {
  const factors = scenarioFactors(request.scenario_multiplier);
  const discount = request.acquisition.target_discount_pct * factors.discount_factor;
  const price = request.acquisition.asking_price * (1 - discount);
  return roundMoney(Math.max(price, 0));
}

function notaryAndAcquisitionCosts(purchasePrice: number, request: NormalizedRequest): number {
  const pct = request.acquisition.notary_and_fees_pct;
  if (request.tax_profile.tax_regime === "registry") {
    const regTax = purchasePrice * request.tax_profile.registration_tax_rate;
    return roundMoney(purchasePrice * pct + regTax);
  }
  const vat = purchasePrice * request.tax_profile.vat_rate;
  return roundMoney(purchasePrice * pct + vat);
}

function renovationTotal(request: NormalizedRequest): number {
  const factors = scenarioFactors(request.scenario_multiplier);
  const base = request.renovation.total_capex * factors.capex_factor;
  const contingency = base * request.renovation.contingency_pct;
  return roundMoney(base + contingency);
}

function totalProjectCost(purchase: number, acquisitionCosts: number, capex: number): number {
  return roundMoney(purchase + acquisitionCosts + capex);
}

function initialCapital(totalCost: number, loanAmount: number): number {
  return roundMoney(Math.max(totalCost - loanAmount, 0));
}

function computeLtv(loanAmount: number, propertyValue: number): number | null {
  if (propertyValue <= 0) return null;
  return roundRate(loanAmount / propertyValue);
}

function annualDebtService(loanAmount: number, rateAnnual: number, termYears: number): number {
  if (loanAmount <= 0) return 0;
  const monthlyRate = rateAnnual / 12;
  const n = termYears * 12;
  if (monthlyRate === 0) {
    return roundMoney((loanAmount / n) * 12);
  }
  const factor = (1 + monthlyRate) ** n;
  const monthlyPayment = (loanAmount * monthlyRate * factor) / (factor - 1);
  return roundMoney(monthlyPayment * 12);
}

function computeDscr(netOperatingIncome: number, debtService: number): number | null {
  if (debtService <= 0) return null;
  return roundRate(netOperatingIncome / debtService);
}

function corporateTax(amount: number, request: NormalizedRequest): number {
  if (amount <= 0) return 0;
  const ires = amount * request.tax_profile.ires_rate;
  const irap = amount * request.tax_profile.irap_rate;
  return roundMoney(ires + irap);
}

function annualNetRentalIncome(request: NormalizedRequest): number {
  const gross = request.rental.monthly_rent * 12;
  const vacancy = gross * request.rental.vacancy_rate;
  const maintenance = gross * request.rental.annual_maintenance_pct;
  const management = gross * request.rental.property_management_pct;
  const net = gross - vacancy - maintenance - management;
  return roundMoney(Math.max(net, 0));
}

function saleMargins(
  salePrice: number,
  totalCost: number,
  request: NormalizedRequest
): [number, number] {
  const saleCosts = salePrice * request.sale.sale_costs_pct;
  const gross = roundMoney(salePrice - totalCost - saleCosts);
  const tax = gross > 0 ? corporateTax(gross, request) : 0;
  const net = roundMoney(gross - tax);
  return [gross, net];
}

function computeNpv(cashFlows: number[], discountRateAnnual: number): number {
  const monthlyRate = discountRateAnnual / 12;
  let npv = 0;
  for (let month = 0; month < cashFlows.length; month += 1) {
    const discount = (1 + monthlyRate) ** month;
    npv += cashFlows[month] / discount;
  }
  return roundMoney(npv);
}

function computeIrr(cashFlows: number[], maxIterations = 100): number | null {
  if (!cashFlows.length || cashFlows[0] >= 0) return null;

  let low = -0.99;
  let high = 5;
  let mid = 0;
  for (let i = 0; i < maxIterations; i += 1) {
    mid = (low + high) / 2;
    let npv = 0;
    for (let t = 0; t < cashFlows.length; t += 1) {
      npv += cashFlows[t] / (1 + mid) ** t;
    }
    if (Math.abs(npv) < 0.01) {
      return roundRate(mid * 12);
    }
    if (npv > 0) {
      low = mid;
    } else {
      high = mid;
    }
  }
  return roundRate(mid * 12);
}

function buildCashFlows(
  request: NormalizedRequest,
  purchase: number,
  acquisitionCosts: number,
  capex: number,
  strategy: DealStrategy
): number[] {
  const months = Math.max(request.timeline.exit_month, request.sale.holding_months, 12);
  const flows = new Array<number>(months + 1).fill(0);

  const equityOut = initialCapital(
    totalProjectCost(purchase, acquisitionCosts, capex),
    request.financing.loan_amount
  );
  flows[0] = -equityOut;
  flows[request.timeline.acquisition_month] -= purchase + acquisitionCosts;

  const renoStart = request.timeline.renovation_start_month;
  const renoMonths = Math.max(request.renovation.duration_months, 1);
  const capexMonthly = capex / renoMonths;
  for (let m = renoStart; m < Math.min(renoStart + renoMonths, months + 1); m += 1) {
    flows[m] -= capexMonthly;
  }

  const debtMonthly = annualDebtService(
    request.financing.loan_amount,
    request.financing.interest_rate_annual,
    request.financing.loan_term_years
  ) / 12;

  if (strategy === "buy_renovate_rent" || strategy === "buy_hold_sell") {
    const netRent = annualNetRentalIncome(request) / 12;
    for (let m = renoStart + renoMonths; m < months + 1; m += 1) {
      flows[m] += netRent - debtMonthly;
    }
  } else {
    for (let m = 1; m < months + 1; m += 1) {
      flows[m] -= debtMonthly;
    }
  }

  if (strategy === "fix_flip" || strategy === "buy_hold_sell") {
    const exitM = Math.min(request.timeline.exit_month, months);
    const salePrice = adjustedSalePrice(
      request.sale.expected_sale_price,
      request.scenario_multiplier
    );
    const saleCosts = salePrice * request.sale.sale_costs_pct;
    const loanPayoff = request.financing.loan_amount;
    flows[exitM] += salePrice - saleCosts - loanPayoff;
  }

  return flows.map(roundMoney);
}

function sensitivitySignal(
  netMargin: number | null,
  dscr: number | null,
  irr: number | null,
  scenarioMultiplier: number
): SensitivitySignal {
  let score = 0;
  if (netMargin !== null) {
    if (netMargin > 50000) score += 2;
    else if (netMargin > 15000) score += 1;
    else if (netMargin < 0) score -= 2;
  }
  if (dscr !== null) {
    if (dscr >= 1.25) score += 1;
    else if (dscr < 1) score -= 2;
  }
  if (irr !== null) {
    if (irr >= 0.15) score += 1;
    else if (irr < 0.05) score -= 1;
  }
  if (scenarioMultiplier > 1) score -= 1;
  if (score >= 2) return "green";
  if (score >= 0) return "amber";
  return "red";
}

function runScenario(input: AnalysisRequestInput, multiplier?: number): ScenarioResult {
  const request = normalizeRequest(input, multiplier);
  const purchase = acquisitionPrice(request);
  const acqCosts = notaryAndAcquisitionCosts(purchase, request);
  const capex = renovationTotal(request);
  const totalCost = totalProjectCost(purchase, acqCosts, capex);
  const equity = initialCapital(totalCost, request.financing.loan_amount);

  const propValue =
    request.financing.property_value_for_ltv ??
    (request.sale.expected_sale_price || purchase);
  const ltv = computeLtv(request.financing.loan_amount, propValue);

  let grossMargin: number | null = null;
  let netMargin: number | null = null;
  if (request.strategy === "fix_flip" || request.strategy === "buy_hold_sell") {
    const salePrice = adjustedSalePrice(
      request.sale.expected_sale_price,
      request.scenario_multiplier
    );
    [grossMargin, netMargin] = saleMargins(salePrice, totalCost, request);
  }

  let netRent: number | null = null;
  let monthlyCf: number | null = null;
  if (request.strategy === "buy_renovate_rent" || request.strategy === "buy_hold_sell") {
    netRent = annualNetRentalIncome(request);
    const debt = annualDebtService(
      request.financing.loan_amount,
      request.financing.interest_rate_annual,
      request.financing.loan_term_years
    );
    monthlyCf = roundMoney((netRent - debt) / 12);
  }

  const debtService = annualDebtService(
    request.financing.loan_amount,
    request.financing.interest_rate_annual,
    request.financing.loan_term_years
  );
  const dscr = netRent !== null ? computeDscr(netRent, debtService) : null;

  const flows = buildCashFlows(request, purchase, acqCosts, capex, request.strategy);
  const npv = computeNpv(flows, request.timeline.discount_rate_annual);
  const irr = computeIrr(flows);
  const signal = sensitivitySignal(netMargin, dscr, irr, request.scenario_multiplier);

  return {
    initial_capital_required: equity,
    total_project_cost: totalCost,
    gross_sale_margin: grossMargin,
    net_sale_margin: netMargin,
    annual_net_rental_income: netRent,
    monthly_cash_flow: monthlyCf,
    ltv,
    dscr,
    npv,
    irr,
    sensitivity_signal: signal,
    assumptions_used: {
      purchase_price: String(purchase),
      acquisition_costs: String(acqCosts),
      capex: String(capex),
      total_cost: String(totalCost),
      scenario_multiplier: String(request.scenario_multiplier),
    },
  };
}

export function runFullAnalysis(input: AnalysisRequestInput): AnalysisResponse {
  const base = runScenario(input, 1);
  const prudent = runScenario(input, 0.85);
  const stress = runScenario(input, 1.25);

  const signals = [base.sensitivity_signal, prudent.sensitivity_signal, stress.sensitivity_signal];
  let summary: string;
  if (signals.every((s) => s === "green")) {
    summary =
      "Operazione solida su tutti gli scenari. Margine regge anche in caso prudente.";
  } else if (stress.sensitivity_signal === "red") {
    summary =
      "Attenzione: lo scenario stress è in rosso. Servono maggiore sconto, capex contenuto o tempi più rapidi.";
  } else {
    summary =
      "Operazione fattibile nel caso base ma sensibile a capex e tempi. Valutare negoziazione aggiuntiva.";
  }

  return {
    base_case: base,
    prudent_case: prudent,
    stress_case: stress,
    sensitivity_summary: summary,
  };
}
